### DISCUSSION TITLE
How to Write the Video from the numpy array as well as adding the Audio to that Video File.

### DISCUSSION THREAD

**Discussion by SohamTilekar** [python-ffmpegio/python-ffmpegio#21] [Q&A]
> I Read the Documentation but I do not understand it Completely.

I Have a Video clip as Numpy array in memory and An Audio Saved on the Disk. I Want to Write the Video Using the Numpy array With the Audio which is saved on the Disk.

> I Tried to Add the parameter -i in the ffmpeg options. And More

Below is My Current Code: - 

```python
        video_np = np.asarray(tuple(self.iterate_frames_array_t(fps if fps else self.fps if self.fps else (_ for _ in ()
                                                                                                            ).throw(Exception('Make Frame is Not Set.'))))) # Throwing Exception if fps is Not Set

        audio_name, _ = os.path.splitext(filename)

        with tempfile.NamedTemporaryFile(suffix=".wav", prefix=audio_name + "_temp_audio_") as temp_audio_file:
            if self.audio and audio:
                self.audio.write_audio_file(temp_audio_file)
                audio_file_name = temp_audio_file.name

        ffmpeg_options = {
            'preset': preset,
            
            **(ffmpeg_params if ffmpeg_params is not None else {}),
            **({'c:v': codec} if codec else {}),
            **({'b:v': bitrate} if bitrate else {}),
            **({'pix_fmt': pixel_format} if pixel_format else {}),
            **({'c:a': audio_codec} if audio_codec else {}),
            **({'ar': audio_fps} if audio_fps else {}),
            **({'b:a': audio_bitrate} if audio_bitrate else {}),
            **({'threads': threads} if threads else {}),
        } # Only Providing the Parameters if they are Provided in the function

        ffmpegio.video.write(filename, 
                            fps if fps else self.fps if self.fps else (_ for _ in ()).throw(Exception('Make Frame is Not Set.')),
                            video_np,
                            overwrite=over_write_output,
                            show_log=True,
                            **ffmpeg_options)
```

**Comment by tikuma-lsuhsc** [comment]
@SohamTilekar - A timely question as I just worked on the exact same case! Based on your code snippet, using the video stream writer works the best for your case. Try this:

```python
with tempfile.NamedTemporaryFile(
    suffix=".wav", prefix=audio_name + "_temp_audio_"
) as temp_audio_file:
    if self.audio and audio:
        self.audio.write_audio_file(temp_audio_file)
        audio_file_name = temp_audio_file.name

        ffmpeg_options.update(
            {"extra_inputs": [temp_audio_file], "acodec": "copy", "map": ["0:v", "1:a"]}
        )

    with ffmpegio.open(
        filename,
        "wv",
        fps,
        **ffmpeg_options,
        overwrite=over_write_output,
        show_log=True,
    ) as f:
        for I in self.iterate_frames_array_t(fps):
            f.write(I)
```

The `ffmpeg_options.update(...)` line is critical to add the audio file as the second input (the first being the `stdin`). Change `acodec` option if you're encoding it and add additional audio options here as needed. The `map` option may not be needed but just to be on the safe side.

As for the stream writer, it works like the batch writer, but you define the mode `wv` as in "write video". Then, run a for loop to iterate frame-by-frame. I assumed that `self.iterate_frames_array_t()` returns a 2D (if grayscale) or 3D (rgb) data.

If you cannot get it to work, post the ffmpeg log along with your code.

**Reply by SohamTilekar** [reply]
I Am Able to Represent the Audio Clip as the Numpy Array Also.

**Reply by SohamTilekar** [reply]
I Get an Error: -

> ```
> Traceback (most recent call last):
>   File "d:\soham_code\video_py\test.py", line 5, in <module>
>     clip.write_videofile(r'D:\soham_code\video_py\test\test_video.mp4')
>   File "d:\soham_code\video_py\vidiopy\video\VideoClips.py", line 190, in write_videofile
>     with ffmpegio.open(
>   File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python312\Lib\contextlib.py", line 137, in __enter__
>     return next(self.gen)
>            ^^^^^^^^^^^^^^
>   File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python312\Lib\site-packages\ffmpegio\__init__.py", line 324, in open
>     stream = StreamClass(*args, **kwds)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python312\Lib\site-packages\ffmpegio\streams\SimpleStreams.py", line 489, in __init__
>     super().__init__(
>   File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python312\Lib\site-packages\ffmpegio\streams\SimpleStreams.py", line 365, in __init__
>     configure.add_url(ffmpeg_args, "input", *input)
> TypeError: add_url() takes from 3 to 5 positional arguments but 765 were given
> ```

Modified Code: -
```
    def write_videofile(self, filename, fps=None, codec=None,   
                        bitrate=None, audio=True, audio_fps=44100,
                        preset="medium", pixel_format=None,
                        audio_codec=None, audio_bitrate=None,
                        write_logfile=False, verbose=True,
                        threads=None, ffmpeg_params: dict[str, str] | None = None,
                        logger='bar', over_write_output=True):
        
        # video_np = np.asarray(tuple(self.iterate_frames_array_t(fps if fps else self.fps if self.fps else (_ for _ in ()
        #                                                                                                     ).throw(Exception('Make Frame is Not Set.')))))

        audio_name, _ = os.path.splitext(filename)


        ffmpeg_options = {
            'preset': preset,
            
            **(ffmpeg_params if ffmpeg_params is not None else {}),
            **({'c:v': codec} if codec else {}),
            **({'b:v': bitrate} if bitrate else {}),
            **({'pix_fmt': pixel_format} if pixel_format else {}),
            **({'c:a': audio_codec} if audio_codec else {}),
            **({'ar': audio_fps} if audio_fps else {}),
            **({'b:a': audio_bitrate} if audio_bitrate else {}),
            **({'threads': threads} if threads else {}),
        }

        with tempfile.NamedTemporaryFile(
            suffix=".wav", prefix=audio_name + "_temp_audio_"
        ) as temp_audio_file:
            if self.audio and audio:
                self.audio.write_audio_file(temp_audio_file)
                # audio_file_name = temp_audio_file.name

                ffmpeg_options.update(
                    {"extra_inputs": [temp_audio_file], "acodec": "copy", "map": ["0:v", "1:a"]}
                )

            with ffmpegio.open(
                filename,
                "wv",
                fps,
                **ffmpeg_options,
                overwrite=over_write_output,
                show_log=True,) as f:
                
                for I in self.iterate_frames_array_t(fps):  # if fps else self.fps if self.fps else (_ for _ in ()
                #                                                                                             ).throw(Exception('Make Frame is Not Set.'))):
                    f.write(I)
```

**Comment by SohamTilekar** [comment]
I Modified the code Little bit Then I Get the BrokenPipe Error.

Modified Code: -
```

    def write_videofile(self, filename, fps=None, codec=None,   
                        bitrate=None, audio=True, audio_fps=44100,
                        preset="medium", pixel_format=None,
                        audio_codec=None, audio_bitrate=None,
                        write_logfile=False, verbose=True,
                        threads=None, ffmpeg_params: dict[str, str] | None = None,
                        logger='bar', over_write_output=True):
        
        video_np = np.asarray(tuple(self.iterate_frames_array_t(fps if fps else self.fps if self.fps else (_ for _ in ()
                                                                                                                         ).throw(Exception('Make Frame is Not Set.')))))

        audio_name, _ = os.path.splitext(filename)


        ffmpeg_options = {
            'preset': preset,
            
            **(ffmpeg_params if ffmpeg_params is not None else {}),
            **({'c:v': codec} if codec else {}),
            **({'b:v': bitrate} if bitrate else {}),
            **({'pix_fmt': pixel_format} if pixel_format else {}),
            **({'c:a': audio_codec} if audio_codec else {}),
            **({'ar': audio_fps} if audio_fps else {}),
            **({'b:a': audio_bitrate} if audio_bitrate else {}),
            **({'threads': threads} if threads else {}),
        }

        with tempfile.NamedTemporaryFile(
            suffix=".wav", prefix=audio_name + "_temp_audio_"
        ) as temp_audio_file:
            if self.audio and audio:
                self.audio.write_audio_file(temp_audio_file)

                ffmpeg_options.update(
                    {"extra_inputs": [temp_audio_file.name], "acodec": "copy", "map": ["0:v", "1:a"]}
                )

            with ffmpegio.open(
                filename,
                "wv",
                fps,
                **ffmpeg_options,
                overwrite=over_write_output,
                show_log=True,) as f:

                for I in video_np:
                    f.write(I)
```


Error: -

> ```
> Trailing option(s) found in the command: may be ignored.
> fd:: Invalid data found when processing input
> Traceback (most recent call last):
>   File "d:\soham_code\video_py\test.py", line 5, in <module>
>     clip.write_videofile(r'D:\soham_code\video_py\test\test_video.mp4')
>   File "d:\soham_code\video_py\vidiopy\video\VideoClips.py", line 198, in write_videofile
>     f.write(I)
>   File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python312\Lib\site-packages\ffmpegio\streams\SimpleStreams.py", line 460, in write
>     raise e
>   File "C:\Users\Lenovo\AppData\Local\Programs\Python\Python312\Lib\site-packages\ffmpegio\streams\SimpleStreams.py", line 457, in write
>     self._proc.stdin.write(self._viewer(obj=data))
> BrokenPipeError: [Errno 32] Broken pipe
> ```

**Reply by SohamTilekar** [reply]
It Work Normally But can't add Audio.
```
        ffmpegio.video.write(filename, 
                             fps if fps else self.fps if self.fps else None,
                             video_np, show_log=True, overwrite=over_write_output)
```

**Reply by tikuma-lsuhsc** [reply]
`extra_inputs` argument is only available for stream writer only.  

Is `show_log=True` not printing FFmpeg log? If yes, please post the full log. Also, what are the `dtype` and `shape` of `video_np`?

You can also create video-only file then run 

```python
ffmpegio.transcode([videofile,audiofile], outfile, acodec='copy', vcodec='copy', map=['0:v','1:a'])
```

**Reply by SohamTilekar** [reply]
Dose the video Has Audio File Attached.

**Reply by SohamTilekar** [reply]
The show_log=True Do not Show Log.

Below is all thigging You Required.

```
video_np=array([...], dtype=uint8),

 video_np.dtype=dtype('uint8'), video_np.shape=(242, 360, 640, 3), video_np.size=167270400, video_np.ndim=4
```

**Reply by tikuma-lsuhsc** [reply]
> Dose the video Has Audio File Attached.

Is this referring to my transcode example line?

If so, this is what I meant:

```python
ffmpegio.video.write(temp_video_file, 
                             fps if fps else self.fps if self.fps else None,
                             video_np, show_log=True, overwrite=over_write_output)
ffmpegio.transcode([temp_videofile,temp_audio_file], filename, acodec='copy', vcodec='copy', map=['0:v','1:a'])
```

`temp_video_file` is a temporary file like you're already doing with the audio stream.

> The show_log=True Do not Show Log.

Your package's logging environment must be blocking it then. I double-checked with the latest version and it does print the log. I suggest you to create a minimal example script first before putting it inside your package.

**Comment by SohamTilekar** [comment]
If You Want to see whole Code The Repo Link: - [vidiopy](https://github.com/SohamTilekar/vidiopy).

How to use: -
```
# file test.py in the Cloned Repo.
from vidiopy.editor import VideoFileClip

from moviepy.editor import concatenate_audioclips, CompositeAudioClip

if __name__ == '__main__':
    clip = VideoFileClip(r'D:\soham_code\video_py\test\chaplin.mp4')
    clip.write_videofile(r'D:\soham_code\video_py\test\test_video.mp4')
```

**Reply by tikuma-lsuhsc** [reply]
For your use case, I recommend you to manage and invoke the ffmpeg calls directly with `subprocess.run` or `subprocess.Popen`. It is pretty straight forward and faster. Dependency on `ffmpegio` also unnecessarily bloats to your package.

If you want to see what your `ffmpegio` function call result in ffmpeg command line, turn on your `logging` to DEBUG and it'll print the full command line arguments. To pass a Numpy array to stdin, convert it to `memoryview` and write it to the stream.

**Reply by SohamTilekar** [reply]
Finaly Is There any Way to Run FFMPEG Command from string or (Non Dict) But I want to Get the progress CallBack.

**Reply by tikuma-lsuhsc** [reply]
Study [`ffmpegio/ffmpegprocess.py`](https://github.com/python-ffmpegio/python-ffmpegio/blob/ef9d744c8ac1414083a516e228cdf0429281325d/src/ffmpegio/ffmpegprocess.py). Specifically [`monitor_progress()`](https://github.com/python-ffmpegio/python-ffmpegio/blob/ef9d744c8ac1414083a516e228cdf0429281325d/src/ffmpegio/ffmpegprocess.py#L188) and how it's being used in [`Popen`](https://github.com/python-ffmpegio/python-ffmpegio/blob/ef9d744c8ac1414083a516e228cdf0429281325d/src/ffmpegio/ffmpegprocess.py#L188).

Be aware that FFmpeg does not tell you the percentage completion. If you need that, you need to compute it based on the total number of frames that you're processing.
